heron/cgi-bin/heron-wsgi.py

```diff
@@ -0,0 +1,397 @@
+            shutil.rmtree(work_dir)
+            raise
+
+        shutil.rmtree(work_dir)
+
+    status = '200 OK'
+    response_headers = [   
+        ('Content-type', mime),
+        ("Content-Disposition","attachment; filename=%s" % filename),
+        ("Content-Title", filename),
+        ("Content-Length", data_len)
+    ]
+    return (status, response_headers, data)
+
+
+# Echo uploaded file back to client as data.
+def upload():
+    if not param_available(['mime', 'file']):
+        return
+
+    # Get the form-based data values
+    mime = params.getvalue('mime')
+    file_item = params['file']
+    encoding = params.getvalue('encoding', 'none')
+
+    # Test if the file was uploaded
+    if file_item.filename:
+        # strip leading path from file name to avoid directory traversal attacks
+        # fn = os.path.basename(fileitem.filename)
+        # open('files/' + fn, 'wb').write(fileitem.file.read())
+        # Echo back file content to client
+        # print tempfile.gettempprefix()
+        # print file_item.filename
+        # file_path = os.path.join(self.path, file_item.filename)
+        # temp_file = tempfile.TemporaryFile()
+        # file_path = file_item.file.name
+        data = file_item.value
+
+        # if the upload is a .zip file we assume a zipped ESRI Shapefile
+        # we convert it to GeoJSON, such that the client can read it
+        # The config in the Heron client (Upload or Editor) should then have an entry like:
+        # {name: 'ESRI Shapefile (1 laag, gezipped)', fileExt: '.zip', mimeType: 'text/plain', formatter: 'OpenLayers.Format.GeoJSON'}
+        f, file_ext_in = os.path.splitext(file_item.filename.lower())
+        if file_ext_in == '.zip' or file_ext_in == '.gpkg' or file_ext_in == '.csv' or 'target_srs' in params:
+            # Convert with ogr2ogr
+            work_dir = prepare_dir(suffix='_upwrk')
+
+            in_file, out_file = prepare_upload_files(file_item, file_ext_in, work_dir)
+            assign_srs = params.getvalue('assign_srs', None)
+            source_srs = params.getvalue('source_srs', None)
+            target_srs = params.getvalue('target_srs', None)
+
+            out_file = ogr2ogr(out_file, in_file, 'GeoJSON', assign_srs=assign_srs, source_srs=source_srs, target_srs=target_srs)
+            data = get_file_data(out_file)
+            shutil.rmtree(work_dir)
+
+        if encoding == 'base64':
+            data = base64.b64encode(data)
+        elif encoding == 'url':
+            data = urllib.quote(data)
+        elif encoding == 'escape':
+            data = cgi.escape(data)
+    else:
+        data = 'No file data received'
+
+    status = '200 OK'
+    response_headers = [('Content-type', mime),("Content-Length", len(data))]
+    return (status, response_headers, data)
+
+# Action handlers: jump table with function pointers
+HANDLERS = {
+    'download': download,
+    'upload': upload
+}
+
+def application(environ, start_response):
+    global params
```

Comment:
It would be neater to get rid of this global. But I wanted to keep things in line with heron.cgi